### PR TITLE
API CHANGE: Allow for launch cleanup

### PR DESCRIPTION
**Pull request by TheRealFalcon**
```
API CHANGE: Allow for launch cleanup

This commit removes the `wait` parameter from launch calls and allows
all instance classes to be used as context managers.

When calling `launch(wait=True)`, callers have no mechanism to cleanup
instances that fail on launch because a handle to the created instance
hasn't been returned yet. To remedy this, always return the created
instance immediately and force the caller to call wait() manually.

Additionally, instance classes have been converted into context
managers so that instances can be launched with cleanup being
handled automatically.

Fixes: #262
```

### CODE REVIEW

**Comment by TheRealFalcon** [comment]
This is one possible approach of fixing #262 . Any feedback is welcome!

File: pycloudlib/cloud.py
Comment (blackboxsw) [inline pycloudlib/cloud.py:140]: Drop wait from docstr too

File: pycloudlib/instance.py
Comment (blackboxsw) [inline pycloudlib/instance.py:48]: This needs try/except: pass gating as there may be integration tests that already manually delete and instance before exiting the context mgr.

**Review by blackboxsw — COMMENTED** [review]
This looks good for handling instances specifically. What do you think about using the BaseCloud itself as a contextmanager?
We could persist/append any created resource to  BaseCloud.resources and on __exit__ from the BaseCloud instance we walk self.resources calling `delete` if available?
That'd allow us to track any VPC, resource groups that pycloudlib creates on our behalf as a side-effect of cloud.launch and allow removal of those resources when the test is over.

File: pycloudlib/instance.py
Comment (blackboxsw) [reply pycloudlib/instance.py:48]: flake will tell you you need docstrs on these

File: tests/integration_tests/test_public_api.py
Comment (blackboxsw) [inline tests/integration_tests/test_public_api.py:138]: This and the `finally:  cloud.delete_image(snapshot_id)` above could be automatically removed if our BaseCloud were used a context manager that collected any resources created during that test run and purged those resources on `__exit__` from the context.  We'd probably want to also create a public method `BaseCloud.release_resources`  or `delete_resources` or something to allow folks not using a context manager to manually invoke this at the end of any tests they are running anyway.

**Comment by TheRealFalcon** [comment]
This grew to be a lot bigger than I originally expected...
```
1. Fix unit tests
2. Add a 'clean' method/context manager to cloud API that cleans up all created resources during life of the instance.
3. Ensure exceptions aren't thrown when cloud.clean()/instance.delete() are called multiple times for the same instance.
4. Collect any exceptions raised during clean/delete and return them to caller. This is so that an exception can't interrupt other cleanup that would otherwise happen. Context managers will print these to stderr.
```

Note that this is still untested on openstack as I need to get my openstack creds back before I can test there.

File: pycloudlib/ec2/instance.py
Comment (blackboxsw) [inline pycloudlib/ec2/instance.py:396]:
```suggestion
        # Wait for detach of all NICs matching the provided ip_address
```

File: pycloudlib/ec2/instance.py
Comment (blackboxsw) [inline pycloudlib/ec2/instance.py:389]: Is the reason we aren't realizing the full list of `nics_matching_ips` outside of the for loop below because we are expecting our self._instance.reload() to show the dwindling list of matching NICs over a period of loop iterations?

File: pycloudlib/ec2/instance.py
Comment (blackboxsw) [inline pycloudlib/ec2/instance.py:258]: Doesn't  this [`DeleteOnTermination`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2/client/modify_network_interface_attribute.html#EC2.Client.modify_network_interface_attribute) mean we don't have to worry about memoizing created_interfaces on AWS?

File: pycloudlib/instance.py
Comment (blackboxsw) [inline pycloudlib/instance.py:51]: Feels like we do not want a sys.exit(1) on each instance deletion in the cloud, but probably something we need to track when `cloud.clean` is called though.

File: pycloudlib/azure/cloud.py
Comment (TheRealFalcon) [reply pycloudlib/azure/cloud.py:963]: 1. Probably. I wasn't aware of a delete poller, so I'll look into that.
2. On line 956, we only use self.resource_group if we didn't get a resource group name passed in. The clean method passes over our created resource groups and does that.

File: pycloudlib/azure/cloud.py
Comment (TheRealFalcon) [reply pycloudlib/azure/cloud.py:974]: Some of this is a bit tacked on as I didn't want to do broader class refactors at the same time. If you look at `_create_resource_group()`, if the resource group under the given tag already exists, we'll use it, otherwise we'll create it and add it to `self.created_resource_groups` while also making it `self.resource_group`, so I don't think clean will do nothing here unless the resource group already existed.

File: pycloudlib/cloud.py
Comment (TheRealFalcon) [reply pycloudlib/cloud.py:84]: Good question. My initial thinking is I don't think clean failures need to be catastrophic, but I could also see the case where something is broken and you don't want to keep going. I think it'd be more common that a cleanup failure is something minor, like timeout before resource is gone, so I think it's better to continue rather than exit.

You have any strong opinions?

File: pycloudlib/ec2/cloud.py
Comment (TheRealFalcon) [reply pycloudlib/ec2/cloud.py:280]: I kept going back and forth on whether to log or not. I started out by logging and then decided to not, but missed this case. I can remove it, but can also add other logs if you think it is valuable.

File: pycloudlib/ec2/instance.py
Comment (TheRealFalcon) [reply pycloudlib/ec2/instance.py:258]: You'd think so, but I kept seeing NICs sticking around. I put a comment about it in the `delete()` method.

File: pycloudlib/ec2/instance.py
Comment (TheRealFalcon) [reply pycloudlib/ec2/instance.py:389]: We're not expecting a single IP to be attached to multiple NICs. I'm not sure if that's even possible, but the way the add_network_interface() code works, it should be 1:1. We loop through all of the nics (via list comprehension) because that's the only way to know if we have a nic that matches this particular IP, but we're only ever expecting 1 result from the loop.

Does that make sense? Do you think there's a way to make it clearer?

File: pycloudlib/instance.py
Comment (TheRealFalcon) [reply pycloudlib/instance.py:51]: Track how?

**Comment by TheRealFalcon** [comment]
@blackboxsw I believe I addressed all of your comments that don't require more discussion.

File: pycloudlib/cloud.py
Comment (blackboxsw) [reply pycloudlib/cloud.py:84]: I think we definitely should exit(1) here. As BEnjamin's comment on the original issue was:
` Regarding overall verbosity of different implementations, I care less about number of lines, and care more about the chance of accidentally leaking / orphaning cloud resources, especially with default parameters.`

File: pycloudlib/ec2/instance.py
Comment (blackboxsw) [reply pycloudlib/ec2/instance.py:388]: Nah, it's just the same boilerplate code used a few lines below. Maybe separate that to a private function `_get_nic_for_ip` and call it here and within the for loop?

File: pycloudlib/instance.py
Comment (blackboxsw) [reply pycloudlib/instance.py:51]: Sorry. 'track' is a bad word. I want to be noisy if we fail on instance or image deletion or disk deletion as those are costly things to keep around on some clouds.  The [sys.exit(1) was the intent of my comment above to "track" by raising an error via non-zero exit in pycloudlib/cloud.py on __exit__ if teardown/deletion of some resources failed at the end of a test suite](https://github.com/canonical/pycloudlib/pull/305/files#r1248110230).

File: pycloudlib/cloud.py
Comment (TheRealFalcon) [reply pycloudlib/cloud.py:84]: Rather than exit(1), I think we raise a separate custom exception instead. Something like `CleanupException`.

exit(1) is a pretty broad hammer that gives nobody the option to recover.

Sound good?

Comment (blackboxsw) [reply pycloudlib/cloud.py:84]: Much better.

**Comment by TheRealFalcon** [comment]
@blackboxsw I addressed your comments and rebased the branch.

**Review by blackboxsw — APPROVED** [review]
Perfect, validated failure behavior injecting a zerodiv error in instance.delete on lxd. LGTM

File: pycloudlib/azure/instance.py
Comment (CalvoM) [inline pycloudlib/azure/instance.py:146]: What if we delete the resource group instead of the vm, since it will delete every other resource allocated in the same resource group?

**Comment by TheRealFalcon** [comment]
@hammerstefan or @mjd-x , since you have both been involved in discussions around this in the past, I want you to have a chance to review or raise any concerns before this gets merged. Will this work for CPC?

**Review by mjd-x — APPROVED** [review]
lgtm, thanks! If possible, it'd be nice to see this reflected in the documentation. It should be clear that users are responsible for doing the cleanup for any instance launched, even if errors were raised (for example, even if PycloudlibTimeoutError is raised, the instance could still be running in the cloud)

File: docs/resource_cleanup.md
Comment (blackboxsw) [inline docs/resource_cleanup.md:23]:
```suggestion
Alternatively, if you don't want to use context managers, you can manually cleanup all resources using the `.clean()` method on `Cloud` objects and the `.delete()` method on `Instance` objects. For example, using EC2:
```

**Review by blackboxsw — APPROVED** [review]
minor doc typo. example looks good +1 on this.

**Pull request merged**
